File: src/main/java/analyzer/histories/Issue.java
package analyzer.histories;

import org.json.JSONArray;
import org.json.JSONObject;
import util.ReaderTool;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Issue {
    public String id = "";
    public String title;
    public Map<String, String> details = new HashMap<>();
    public String description;
    public List<String> attachements = new ArrayList<>();
    public Map<String, String> issueLinks = new HashMap<>();
    public List<Comment> comments;

    public Issue(String issueId) {
        String issueContent = ReaderTool.read("issueCrawler/issue/" + issueId + ".json");
        if (issueContent.equals("")) return ;
        JSONObject object = new JSONObject(issueContent);
        this.id = get("id", object);
        this.title = get("title", object);
        this.description = get("description", object);

        JSONObject detailsObject = (JSONObject) object.get("details");
        details.put("type", get("type", detailsObject));
        details.put("status", get("status", detailsObject));
        details.put("priority", get("priority", detailsObject));
        details.put("resolution", get("resolution", detailsObject));
        details.put("affect_versions", get("affect_versions", detailsObject));
        details.put("labels", get("labels", detailsObject));

        //这两个鬼东西没有处理
        //details.put("fix_versions", (String) detailsObject.get("fix_versions"));
        //details.put("components")

        this.comments = new ArrayList<>();
        JSONArray comments = (JSONArray) object.get("comments");

        comments.forEach(comment -> {
            JSONObject com = (JSONObject) comment;
            com.put("issue_id", this.id);
            com.put("issue_title", this.title);
            this.comments.add( new Comment((JSONObject) comment));
        });

        Comment comment = new Comment();
        comment.issueId = this.id;
        comment.issueTitle = this.title;
        comment.content = this.description;
        this.comments.add(comment);
    }

    private String get(String field, JSONObject json) {
        String result = "";
        try{
            result = (String) json.get(field);
            return result;
        }catch (Exception e) {
            return "";
        }
    }

    public static void main(String[] args){
        Issue issue = new Issue("SOLR-1");

    }


}

File: issueCrawler/issue_crawler/Issue.py
from bs4 import BeautifulSoup
import json
import collections


class Issue:
    def __init__(self, page):
        self.status = True
        self.id = ""
        self.title = ""
        self.details = {}
        self.description = ""
        self.attachments = {}
        self.issue_links = {}
        self.comments = []

        self.init_id(page)
        self.init_title(page)
        self.init_details(page)
        self.init_description(page)
        self.init_attachments(page)
        self.init_issue_links(page)
        self.init_comment(page)

    def init_id(self, page):
        self.id = page.find('a', {'id': 'key-val'}).text.strip()

    def init_title(self, page):
        self.title = page.find('h1', {'id': 'summary-val'}).text.strip()

    def init_details(self, page):
        detail = page.find('ul', {'id': 'issuedetails'})

        self.details['type'] = detail.find('span', {'id': 'type-val'}).text.strip()
        self.details['status'] = detail.find('span', {'id': 'status-val'}).text.strip()
        self.details['priority'] = detail.find('span', {'id': 'priority-val'}).text.strip()
        self.details['resolution'] = detail.find('span', {'id': 'resolution-val'}).text.strip()
        self.details['affect_versions'] = detail.find('span', {'id': 'versions-val'}).text.strip()

        self.details['fix_versions'] = []
        for a in detail.find('span', {'id': 'fixfor-val'}).find_all('a'):
            self.details['fix_versions'].append(a.text.strip())

        self.details['components'] = []
        for a in detail.find('span', {'id': 'components-val'}).find_all('a'):
            self.details['components'].append(a.text.strip())

        self.details['labels'] = ""  # detail.find('span', {'class': 'labels'}).text.strip()

    def init_description(self, page):
        des = page.find('div', {'id':'descriptionmodule'})
        if des:
            self.description = des.find('div', {'class': 'mod-content'}).get_text().strip()

    def init_attachments(self, page):
        atta = page.find('ol', {'id': 'file_attachments'})
        if atta:
            for li in atta.find_all('li'):
                dt = li.find('dt', {'class': 'attachment-title'})
                patch = dt.find('a').text.strip()
                url = 'https://issues.apache.org' + dt.find('a')['href'].strip()
                self.attachments[patch] = url

    def init_issue_links(self, page):
        # TODO
        pass

    def init_comment(self, page):
        comment_container = page.find('div', {'id': 'issue_actions_container', 'class': 'issuePanelContainer'})
        if comment_container:
            for div in comment_container.find_all('div', class_='issue-data-block activity-comment twixi-block expanded'):
                comment = dict()
                comment['id'] = div['id']
                if comment['id'] == 'comment-12508577':
                    v = 2
                comment['date'] = div.find('time')['datetime']
                comment['author'] = div.find(['span', 'a'], class_='user-hover user-avatar').text.strip()
                comment['content'] = div.find('div', {'class': 'action-body flooded'}).get_text()
                self.comments.append(comment)

    def toString(self):
        if self.status is True:
            json_object = collections.OrderedDict()
            json_object['id'] = self.id
            json_object['title'] = self.title
            json_object['details'] = self.details
            json_object['description'] = self.description
            json_object['attachments'] = self.attachments
            json_object['issue_links'] = self.issue_links
            json_object['comments'] = self.comments
            s = json.dumps(json_object, indent=4)
            return s
        else:
            return None


if __name__ == '__main__':
    pass


